### PR TITLE
XDS Server: introducing snapshot cache

### PR DESCRIPTION
**Pull request by eduser25**
Go-control-plane includes an XDS server implementation that is interacted
through a map-like mechanisms that helps decouple generating proxy configuration
from its distribution.

https://github.com/envoyproxy/go-control-plane/blob/main/docs/cache/Snapshot.md

This commit adds support for initializing OSM control plane to use such mechanism, behind
a feature flag.

Using the cache currently suffers from some problems (Some e2e work, Some are flacky, most will fail).
More work is to be put on top of this implementation to debug these issues. This commit introduces merely
the foundations to keep things short.

Meaningful changes:
- New featureflag added: `enable-snapshotcache`
- `server.go` will initialize the XDS server with or without the cache when enabled.
- Some changes were made to `ads/response.go` to accomodate both modes working.
Namely, it will now `get` the resources for a vertical, and later decide what to do with
them (either send to a connected proxy, or store them in the cache, depending if the cache
is enabled or not).
- `stream.go` was before the routine receiving events and triggering envoy re-configs. When
working on snapshot-cache, such routines do not exist, so a new routine to mimic the same
current design and behavior has been added (`snapshot_cache_stream.go`). Its purpose is to
listen to broadcast events from dispatcher, and issue a reconfiguration for all proxies when
a broadcast update is sent from dispatcher.

TODOs:
- Proxy objects are not created anymore in snapshot-cache based. Follow up discussion if
we want to have them or not is to happen.
- Zerolog doesn't implement Golang's logging interface and can't be used/passed to log
cache internals.
- Additional context validation (certificate) can happen on cache callback context.
- We have no way to infer the `kind` of proxy from a Pod. Right now it's temporarily
hardcoded as `sidecar`.

Signed-off-by: Eduard Serra <eduser25@gmail.com>

| New Functionality          | [x] |
| Control Plane              | [x] |

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?
No
1. Is this a breaking change?
No

### CODE REVIEW

File: pkg/envoy/ads/snapshot_cache_stream.go
Comment (eduser25) [inline pkg/envoy/ads/snapshot_cache_stream.go:71]: ?????

File: pkg/envoy/ads/response.go
Comment (shashankram) [inline pkg/envoy/ads/response.go:18]:
```suggestion
// getTypeResource invokes the XDS handler (LDS, CDS etc.) to respond to the XDS request containing the requests' type and associated resources
```

File: pkg/envoy/ads/response.go
Comment (shashankram) [inline pkg/envoy/ads/response.go:23]: > proxy with SerialNumber=%s on Pod with UID=%s

Every time I see this log 👀

File: pkg/envoy/ads/response.go
Comment (shashankram) [inline pkg/envoy/ads/response.go:116]: Consider moving this to `ads/cache.go`?

File: pkg/envoy/ads/server.go
Comment (shashankram) [inline pkg/envoy/ads/server.go:57]: nil value default?

File: pkg/envoy/ads/server.go
Comment (shashankram) [inline pkg/envoy/ads/server.go:81]: I got the comment, but correlating it to the code is missing.

File: pkg/envoy/ads/snapshot_cache_stream.go
Comment (eduser25) [reply pkg/envoy/ads/snapshot_cache_stream.go:70]: very likely, let me see

File: cmd/osm-controller/osm-controller.go
Comment (snehachhabria) [inline cmd/osm-controller/osm-controller.go:110]: FYI, we are planning to remove feature flags and move these into the mesh config : https://github.com/openservicemesh/osm/pull/3536 . You might want to rebase this once the former is merged. 

cc @jaellio

File: pkg/envoy/ads/snapshot_cache_stream.go
Comment (snehachhabria) [inline pkg/envoy/ads/snapshot_cache_stream.go:55]: Could we add a unit test for this ?

File: pkg/envoy/ads/response.go
Comment (eduser25) [reply pkg/envoy/ads/response.go:40]: actually... The errors types logged and returned are different. What we return are generic errors while what we log here is more specific. Not sure how to take this, but I think we'd rather change it in a follow up (we should decide if we want specific logs to be returned, generic logs to be logged, etc).

File: pkg/envoy/ads/snapshot_cache_stream.go
Comment (eduser25) [reply pkg/envoy/ads/snapshot_cache_stream.go:55]: writing :+1:

File: cmd/osm-controller/osm-controller.go
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:110]: are we racing? :eyes:

**Review by shashankram — APPROVED** [review]

File: cmd/osm-controller/osm-controller.go
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go]: rebased

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**
